src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java

```diff
@@ -0,0 +1,411 @@
+  private SnowflakeTelemetryPipeStatus pipeTelemetry;
+  private SnowflakeTelemetryService telemetryService;
+
+  private PipeProgressRegistryTelemetry telemetry;
+  private AtomicInteger loops;
+  private AtomicReference<BiConsumer<Integer, Long>> nextTickCallback;
+  private AtomicReference<ScheduledFuture<?>> scheduledFuture;
+
+  @BeforeEach
+  void setUp() {
+    loops = new AtomicInteger(0);
+    currentTime = new AtomicLong(new DateTime(2000, 1, 10, 12, 0, DateTimeZone.UTC).getMillis());
+    nextTickCallback = new AtomicReference<>();
+    scheduledFuture = new AtomicReference<>();
+
+    conn = Mockito.mock(SnowflakeConnectionService.class);
+    ingestionService = Mockito.mock(SnowflakeIngestionService.class);
+    pipeCreation = new SnowflakeTelemetryPipeCreation(TABLE_NAME, STAGE_NAME, PIPE_NAME);
+    pipeTelemetry = Mockito.mock(SnowflakeTelemetryPipeStatus.class);
+    telemetryService = Mockito.mock(SnowflakeTelemetryService.class);
+    telemetry = new PipeProgressRegistryTelemetry(pipeCreation, pipeTelemetry, telemetryService);
+
+    victim =
+        new StageFilesProcessor(
+            PIPE_NAME,
+            TABLE_NAME,
+            STAGE_NAME,
+            PREFIX,
+            conn,
+            ingestionService,
+            pipeTelemetry,
+            telemetryService,
+            createTestScheduler(loops, currentTime, nextTickCallback, scheduledFuture),
+            currentTime::get);
+    register = new StageFilesProcessor.ProgressRegisterImpl(victim);
+  }
+
+  @Test
+  void filesProcessWillTerminateOnStopSignal() {
+    StageFilesProcessor.ProgressRegister client = victim.trackFilesAsync();
+    client.close();
+    verify(scheduledFuture.get(), times(1)).cancel(true);
+  }
+
+  @Test
+  void fileProcessorWillTryToGetInitialStateFromStageExactlyOnce() {
+    // lets simulate 1 hour (60 cleanup cycles)
+    loops.set(60);
+
+    when(conn.listStage(STAGE_NAME, PREFIX)).thenReturn(new ArrayList<>());
+
+    victim.trackFiles(register, telemetry);
+
+    verify(conn, times(1)).listStage(STAGE_NAME, PREFIX);
+  }
+
+  @Test
+  void fileProcessorWillTryGettingStateFromStageOnError() {
+    String ingestFile = String.format("connector/topic/0/1_9_%d.json.gz", currentTime.get());
+    register.registerNewStageFile(ingestFile);
+
+    // lets simulate 10 cleanup cycles
+    loops.set(10);
+
+    when(conn.listStage(STAGE_NAME, PREFIX)).thenReturn(new ArrayList<>());
+    when(ingestionService.readIngestHistoryForward(anyMap(), any(), any(), anyInt()))
+        .thenReturn(0)
+        .thenThrow(new SnowflakeKafkaConnectorException("expected", "error"))
+        .thenReturn(0);
+
+    victim.trackFiles(register, telemetry);
+
+    verify(conn, times(2)).listStage(STAGE_NAME, PREFIX);
+    verify(ingestionService, times(10)).readIngestHistoryForward(anyMap(), any(), any(), anyInt());
+  }
+
+  @Test
+  void fileProcessorWillPurgeLoadedFilesWhenHistoryIsAvailableAfterFilesHaveBeenSubmitted() {
```

Comment:
Nit: it's quite hard to read this test name. Adding underscores might help: `fileProcessor_willPurgeLoadedFiles_whenHistoryIsAvailableAfterFilesHaveBeenSubmitted`.
Applies to the other tests too.